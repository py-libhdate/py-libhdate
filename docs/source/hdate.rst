hdate package
=============

Submodules
----------

hdate.common module
-------------------

.. automodule:: hdate.common
   :members:
   :undoc-members:
   :show-inheritance:

hdate.converters module
-----------------------

.. automodule:: hdate.converters
   :members:
   :undoc-members:
   :show-inheritance:

hdate.date module
-----------------

.. automodule:: hdate.date
   :members:
   :undoc-members:
   :show-inheritance:

hdate.htables module
--------------------

.. automodule:: hdate.htables
   :members:
   :undoc-members:
   :show-inheritance:

hdate.zmanim module
-------------------

.. automodule:: hdate.zmanim
   :members:
   :undoc-members:
   :show-inheritance:

Module contents
---------------

.. automodule:: hdate
   :members:
   :undoc-members:
   :show-inheritance:
