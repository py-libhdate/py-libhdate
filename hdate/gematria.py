"""Gematria for hebrew numbers."""

DIGITS = (
    (" ", "א", "ב", "ג", "ד", "ה", "ו", "ז", "ח", "ט"),
    ("ט", "י", "כ", "ל", "מ", "נ", "ס", "ע", "פ", "צ"),
    (" ", "ק", "ר", "ש", "ת"),
)


def hebrew_number(num: int, language: str = "hebrew", short: bool = False) -> str:
    """Return "Gimatria" number."""
    if language != "hebrew":
        return str(num)
    if not 0 <= num < 10000:
        raise ValueError(f"num must be between 0 to 9999, got:{num}")
    hstring = ""
    if num >= 1000:
        hstring += DIGITS[0][num // 1000]
        hstring += "' "
        num = num % 1000
    while num >= 400:
        hstring += DIGITS[2][4]
        num = num - 400
    if num >= 100:
        hstring += DIGITS[2][num // 100]
        num = num % 100
    if num >= 10:
        if num in [15, 16]:
            num = num - 9
        hstring += DIGITS[1][num // 10]
        num = num % 10
    if num > 0:
        hstring += DIGITS[0][num]
    # possibly add the ' and " to hebrew numbers
    if not short:
        if len(hstring) < 2:
            hstring += "'"
        else:
            hstring = hstring[:-1] + '"' + hstring[-1]
    return hstring
