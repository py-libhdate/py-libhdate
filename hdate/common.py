"""Small helper classes."""

from dataclasses import dataclass
from datetime import tzinfo

import pytz

from hdate.htables import Months


class BaseClass:
    """Implement basic functionality for all classes."""

    def __eq__(self, other):
        """Override equality operator."""
        if isinstance(other, self.__class__):
            return self.__dict__ == other.__dict__
        return False

    def __ne__(self, other):
        """Override inequality operator."""
        return not self.__eq__(other)


@dataclass
class HebrewDate(BaseClass):
    """Define a Hebrew date object."""

    year: int
    month: Months
    day: int

    def __post_init__(self):
        self.month = (
            self.month if isinstance(self.month, Months) else Months(self.month)
        )


class Location(BaseClass):
    """Define a geolocation for Zmanim calculations."""

    # pylint: disable=too-many-arguments
    def __init__(
        self,
        name="Jerusalem",
        latitude=31.778,
        longitude=35.235,
        timezone="Asia/Jerusalem",
        altitude=754,
        diaspora=False,
    ):
        """Initialitze the location object."""
        self._timezone = None
        self.name = name
        self.latitude = latitude
        self.longitude = longitude
        self.timezone = timezone
        self.altitude = altitude
        self.diaspora = diaspora

    def __repr__(self):
        """Return a representation of Location for programmatic use."""
        return (
            f"Location(name='{self.name}', latitude={self.latitude}, "
            f"longitude={self.longitude}, timezone='{self.timezone}', "
            f"altitude={self.altitude}, diaspora={self.diaspora})"
        )

    @property
    def timezone(self):
        """Return the timezone."""
        return self._timezone

    @timezone.setter
    def timezone(self, value):
        """Set the timezone."""
        self._timezone = value if isinstance(value, tzinfo) else pytz.timezone(value)
