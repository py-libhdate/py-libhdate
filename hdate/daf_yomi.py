"""Daf Yomi module."""

import datetime as dt
from dataclasses import dataclass
from typing import ClassVar

from hdate.gematria import hebrew_number
from hdate.translator import TranslatorMixin


@dataclass(frozen=True)
class Masechta(TranslatorMixin):
    """Masechta object."""

    name: str
    pages: int

    def __str__(self) -> str:
        name = self.get_translation(self.name)
        daf = hebrew_number(self.pages + 2, language=self._language, short=True)
        return f"{name} {daf}"


@dataclass
class DafYomiDatabase:
    """Database of Masechtos."""

    # The first few cycles were only 2702 blatt. After that it became 2711. Even with
    # that, the math doesn't play nicely with the dates before the 11th cycle :(
    # From cycle 11 onwards, it was simple and sequential
    _start_date: ClassVar[dt.date] = dt.date(1997, 9, 29)
    _masechtot: ClassVar[list[Masechta]]

    @classmethod
    def register(cls, masechtot: list[Masechta]) -> None:
        """Regisetr a list of masechtos."""
        cls._masechtot = masechtot

    @classmethod
    def cycle_length(cls) -> int:
        """Return the length of a full cycle."""
        return sum(masechta.pages for masechta in cls._masechtot)

    def lookup(self, date: dt.date) -> Masechta:
        """Return the start date of a full cycle."""
        days_since_start = (date - self._start_date).days
        page_number = days_since_start % self.cycle_length()

        masechta_index = 0
        for masechta in self._masechtot:
            if page_number < masechta.pages:
                break
            page_number -= masechta.pages
            masechta_index += 1

        return Masechta(self._masechtot[masechta_index].name, page_number)


DAF_YOMI_MESECHTOS = (
    Masechta("berachos", 63),
    Masechta("shabbos", 156),
    Masechta("eruvin", 104),
    Masechta("pesachim", 120),
    Masechta("shekalim", 21),
    Masechta("yoma", 87),
    Masechta("succah", 55),
    Masechta("beitzah", 39),
    Masechta("rosh_hashanah", 34),
    Masechta("taanis", 30),
    Masechta("megillah", 31),
    Masechta("moed_katan", 28),
    Masechta("chagigah", 26),
    Masechta("yevamos", 121),
    Masechta("kesubos", 111),
    Masechta("nedarim", 90),
    Masechta("nazir", 65),
    Masechta("sotah", 48),
    Masechta("gittin", 89),
    Masechta("kiddushin", 81),
    Masechta("bava_kamma", 118),
    Masechta("bava_metzia", 118),
    Masechta("bava_basra", 175),
    Masechta("sanhedrin", 112),
    Masechta("makkos", 23),
    Masechta("shevuos", 48),
    Masechta("avodah_zarah", 75),
    Masechta("horayos", 13),
    Masechta("zevachim", 119),
    Masechta("menachos", 109),
    Masechta("chullin", 141),
    Masechta("bechoros", 60),
    Masechta("arachin", 33),
    Masechta("temurah", 33),
    Masechta("kereisos", 27),
    Masechta("meilah", 36),
    Masechta("niddah", 72),
)

DafYomiDatabase.register(list(DAF_YOMI_MESECHTOS))
