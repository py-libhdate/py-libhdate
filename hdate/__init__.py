"""
Jewish calendrical date and times for a given location.

HDate calculates and generates a representation either in English or Hebrew
of the Jewish calendrical date and times for a given location
"""
from hdate.common import HebrewDate, Location
from hdate.date import HDate
from hdate.htables import HolidayTypes
from hdate.zmanim import Zmanim

__all__ = ["HDate", "Zmanim", "HebrewDate", "Location", "HolidayTypes"]
