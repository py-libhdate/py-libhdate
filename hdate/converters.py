"""Methods for going back and forth between various calendars."""
import datetime

from hdate.common import HebrewDate
from hdate.htables import Months


def get_chalakim(hours, parts):
    """Return the number of total parts (chalakim)."""
    return (hours * PARTS_IN_HOUR) + parts


PARTS_IN_HOUR = 1080
PARTS_IN_DAY = 24 * PARTS_IN_HOUR
PARTS_IN_WEEK = 7 * PARTS_IN_DAY
PARTS_IN_MONTH = PARTS_IN_DAY + get_chalakim(12, 793)  # Fix for regular month


def _days_from_3744(hebrew_year):
    """Return: Number of days since 3,1,3744."""
    # Start point for calculation is Molad new year 3744 (16BC)
    years_from_3744 = hebrew_year - 3744
    molad_3744 = get_chalakim(1 + 6, 779)  # Molad 3744 + 6 hours in parts

    # Time in months

    # Number of leap months
    leap_months = (years_from_3744 * 7 + 1) // 19
    leap_left = (years_from_3744 * 7 + 1) % 19  # Months left of leap cycle
    months = years_from_3744 * 12 + leap_months  # Total Number of months

    # Time in parts and days
    # Molad This year + Molad 3744 - corrections
    parts = months * PARTS_IN_MONTH + molad_3744
    # 28 days in month + corrections
    days = months * 28 + parts // PARTS_IN_DAY - 2

    # Time left for round date in corrections
    # 28 % 7 = 0 so only corrections counts
    parts_left_in_week = parts % PARTS_IN_WEEK
    parts_left_in_day = parts % PARTS_IN_DAY
    week_day = parts_left_in_week // PARTS_IN_DAY

    # pylint: disable=too-many-boolean-expressions
    # pylint-comment: Splitting the 'if' below might create a bug in case
    # the order is not kept.

    # Molad ד"ר ט"ג
    if (
        (
            leap_left < 12
            and week_day == 3
            and parts_left_in_day >= get_chalakim(9 + 6, 204)
        )
        or
        # Molad ט"פקת ו"טב
        (
            leap_left < 7
            and week_day == 2
            and parts_left_in_day >= get_chalakim(15 + 6, 589)
        )
    ):
        days += 1
        week_day += 1

    # pylint: enable=too-many-boolean-expressions

    # ADU
    if week_day in (1, 4, 6):
        days += 1

    return days


def get_size_of_hebrew_year(hebrew_year):
    """Return: total days in hebrew year."""
    return _days_from_3744(hebrew_year + 1) - _days_from_3744(hebrew_year)


def gdate_to_jdn(date):
    """
    Compute Julian day from Gregorian day, month and year.

    Algorithm from wikipedia's julian_day article.
    Return: The julian day number
    """
    not_jan_or_feb = (14 - date.month) // 12
    year_since_4800bc = date.year + 4800 - not_jan_or_feb
    month_since_4800bc = date.month + 12 * not_jan_or_feb - 3
    jdn = (
        date.day
        + (153 * month_since_4800bc + 2) // 5
        + 365 * year_since_4800bc
        + (year_since_4800bc // 4 - year_since_4800bc // 100 + year_since_4800bc // 400)
        - 32045
    )
    return jdn


def hdate_to_jdn(date):
    """
    Compute Julian day from Hebrew day, month and year.

    Return: julian day number,
            1 of tishrey julians,
            1 of tishrey julians next year
    """
    day = date.day
    month = date.month.value
    if date.month == Months.ADAR_I:
        month = 6
    if date.month == Months.ADAR_II:
        month = 6
        day += 30

    # Calculate days since 1,1,3744
    day = _days_from_3744(date.year) + (59 * (month - 1) + 1) // 2 + day

    # length of year
    length_of_year = get_size_of_hebrew_year(date.year)
    # Special cases for this year
    if length_of_year % 10 > 4 and month > 2:  # long Heshvan
        day += 1
    if length_of_year % 10 < 4 and month > 3:  # short Kislev
        day -= 1
    if length_of_year > 365 and month > 6:  # leap year
        day += 30

    # adjust to julian
    return day + 1715118


def jdn_to_gdate(jdn):
    """
    Convert from the Julian day to the Gregorian day.

    Algorithm from 'Julian and Gregorian Day Numbers' by Peter Meyer.
    Return: day, month, year
    """
    # pylint: disable=invalid-name

    # The algorithm is a verbatim copy from Peter Meyer's article
    # No explanation in the article is given for the variables
    # Hence the exceptions for pylint and for flake8 (E741)

    l = jdn + 68569  # noqa: E741
    n = (4 * l) // 146097
    l = l - (146097 * n + 3) // 4  # noqa: E741
    i = (4000 * (l + 1)) // 1461001  # that's 1,461,001
    l = l - (1461 * i) // 4 + 31  # noqa: E741
    j = (80 * l) // 2447
    day = l - (2447 * j) // 80
    l = j // 11  # noqa: E741
    month = j + 2 - (12 * l)
    year = 100 * (n - 49) + i + l  # that's a lower-case L

    return datetime.date(year, month, day)


def jdn_to_hdate(jdn):
    """Convert from the Julian day to the Hebrew day."""
    # calculate Gregorian date
    date = jdn_to_gdate(jdn)

    # Guess Hebrew year is Gregorian year + 3760
    year = date.year + 3760

    jdn_tishrey1 = hdate_to_jdn(HebrewDate(year, Months.TISHREI, 1))
    jdn_tishrey1_next_year = hdate_to_jdn(HebrewDate(year + 1, Months.TISHREI, 1))

    # Check if computed year was underestimated
    if jdn_tishrey1_next_year <= jdn:
        year = year + 1
        jdn_tishrey1 = jdn_tishrey1_next_year
        jdn_tishrey1_next_year = hdate_to_jdn(HebrewDate(year + 1, Months.TISHREI, 1))

    size_of_year = get_size_of_hebrew_year(year)

    # days into this year, first month 0..29
    days = jdn - jdn_tishrey1

    # last 8 months always have 236 days
    if days >= (size_of_year - 236):  # in last 8 months
        days = days - (size_of_year - 236)
        month = days * 2 // 59
        day = days - (month * 59 + 1) // 2 + 1

        month = month + 4 + 1

        # if leap
        if size_of_year > 355 and month <= 6:
            month = month + 8
    else:  # in 4-5 first months
        # Special cases for this year
        if size_of_year % 10 > 4 and days == 59:  # long Heshvan (day 30)
            month = 1
            day = 30
        elif size_of_year % 10 > 4 and days > 59:  # long Heshvan
            month = (days - 1) * 2 // 59
            day = days - (month * 59 + 1) // 2
        elif size_of_year % 10 < 4 and days > 87:  # short kislev
            month = (days + 1) * 2 // 59
            day = days - (month * 59 + 1) // 2 + 2
        else:  # regular months
            month = days * 2 // 59
            day = days - (month * 59 + 1) // 2 + 1

        month = month + 1

    return HebrewDate(year, Months(month), day)
