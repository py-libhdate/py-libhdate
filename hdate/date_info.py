"""
Hebrew date meta-data.

HDateInfo allows querying various meta-data about Hebrew date, including
Holidays, Daf Yomi, Omer, and more.
"""

from __future__ import annotations

import datetime as dt
from dataclasses import dataclass, field
from typing import Optional, Union

from hdate.daf_yomi import DafYomiDatabase, Masechta
from hdate.gematria import hebrew_number
from hdate.hebrew_date import HebrewDate, Weekday
from hdate.holidays import Holiday, HolidayDatabase, HolidayTypes
from hdate.omer import Omer
from hdate.parasha import Parasha, ParashaDatabase
from hdate.tekufot import Nusachim, Tekufot
from hdate.translator import Language, TranslatorMixin


@dataclass
class HDateInfo(TranslatorMixin):  # pylint: disable=too-many-instance-attributes
    """
    Hebrew date information class.

    Provides access to various properties of a given date.
    """

    date: Union[dt.date, HebrewDate] = field(default_factory=dt.date.today)
    diaspora: bool = False
    language: Language = "hebrew"
    nusach: Nusachim = "sephardi"

    def __post_init__(self) -> None:
        # Initialize private variables
        self._last_updated = ""
        self._holidays = HolidayDatabase(self.diaspora)

        if isinstance(self.date, dt.date):
            self.gdate = self.date
            self._hdate = HebrewDate.from_gdate(self.date)
            self.date = self._hdate  # Replace self.date so comparison always works
        elif isinstance(self.date, HebrewDate):
            self.hdate = self.date
            self._gdate = self.date.to_gdate()
        else:
            raise TypeError("date has to be of type datetime.date or HebrewDate")

        super().__post_init__()

    def __str__(self) -> str:
        in_prefix = "×‘" if self.language == "hebrew" else ""
        day_number = hebrew_number(self.hdate.day, language=self.language)
        year_number = hebrew_number(self.hdate.year, language=self.language)
        result = (
            f"{self.hdate.dow()} "
            f"{day_number} {in_prefix}{self.hdate.month} {year_number}"
        )

        if self.omer:
            result = f"{result} {self.omer}"

        if self.holidays:
            result = f"{result} {', '.join(str(holiday) for holiday in self.holidays)}"
        return result

    @property
    def hdate(self) -> HebrewDate:
        """Return the hebrew date."""
        hdate = (
            self._hdate
            if self._last_updated == "hdate"
            else HebrewDate.from_gdate(self._gdate)
        )
        hdate.set_language(self.language)
        return hdate

    @hdate.setter
    def hdate(self, date: HebrewDate) -> None:
        """Set the dates of the HDateInfo object based on a given Hebrew date."""

        if not isinstance(date, HebrewDate):
            raise TypeError(f"date: {date} is not of type HebrewDate")

        self._last_updated = "hdate"
        self._hdate = date

    @property
    def gdate(self) -> dt.date:
        """Return the Gregorian date for the given Hebrew date."""
        if self._last_updated == "gdate":
            return self._gdate
        return self._hdate.to_gdate()

    @gdate.setter
    def gdate(self, date: dt.date) -> None:
        """Set the Gregorian date for the given Hebrew date."""
        self._last_updated = "gdate"
        self._gdate = date

    @property
    def omer(self) -> Optional[Omer]:
        """Return the Omer object."""
        _omer = Omer(date=self.hdate, language=self.language)
        return _omer if _omer.total_days > 0 else None

    @property
    def parasha(self) -> Parasha:
        """Return the upcoming parasha."""
        db = ParashaDatabase(self.diaspora)
        parasha = db.lookup(self.hdate)
        parasha.set_language(self.language)
        return parasha

    @property
    def holidays(self) -> list[Holiday]:
        """Return the abstract holiday information from holidays table."""
        holidays_list = self._holidays.lookup(self.hdate)

        for holiday in holidays_list:
            holiday.set_language(self.language)

        return holidays_list

    @property
    def daf_yomi(self) -> Masechta:
        """Return the daf yomi for the given date."""
        db = DafYomiDatabase()
        daf = db.lookup(self.gdate)
        daf.set_language(self.language)
        return daf

    @property
    def gevurot_geshamim(self) -> str:
        """Return the rain prayer (Tal uMatar, veTen Beracha, ...)."""
        tekufot = Tekufot(self.gdate, self.diaspora, self.nusach, self.language)
        return tekufot.get_prayer_for_date()

    @property
    def is_shabbat(self) -> bool:
        """Return True if this date is Shabbat.

        Returns False on Friday because the HDate object has no notion of time.
        For more detailed nuance, use the Zmanim object.
        """
        return self.hdate.dow() == Weekday.SATURDAY

    @property
    def is_holiday(self) -> bool:
        """Return True if this date is a holiday (any kind)."""
        return len(self.holidays) > 0

    @property
    def is_yom_tov(self) -> bool:
        """Return True if this date is a Yom Tov."""
        return any(holiday.type == HolidayTypes.YOM_TOV for holiday in self.holidays)

    @property
    def next_day(self) -> HDateInfo:
        """Return the HDateInfo for the next day."""
        return HDateInfo(self.gdate + dt.timedelta(1), self.diaspora, self.language)

    @property
    def previous_day(self) -> HDateInfo:
        """Return the HDateInfo for the previous day."""
        return HDateInfo(self.gdate + dt.timedelta(-1), self.diaspora, self.language)

    @property
    def upcoming_shabbat(self) -> HDateInfo:
        """Return the HDateInfo for either the upcoming or current Shabbat."""
        if self.is_shabbat:
            return self

        next_shabbat = self.gdate + dt.timedelta(Weekday.SATURDAY - self.hdate.dow())
        return HDateInfo(next_shabbat, diaspora=self.diaspora, language=self.language)

    @property
    def upcoming_erev_shabbat(self) -> HDateInfo:
        """Return the HDateInfo for the upcoming or current Erev Shabbat (Friday)."""
        return self.upcoming_shabbat.previous_day

    @property
    def upcoming_yom_tov(self) -> HDateInfo:
        """Return the HDateInfo for the upcoming or current Yom Tov."""
        if self.is_yom_tov:
            return self

        date = self._holidays.lookup_next_holiday(self.hdate, [HolidayTypes.YOM_TOV])

        return HDateInfo(date, self.diaspora, self.language)

    @property
    def upcoming_erev_yom_tov(self) -> HDateInfo:
        """Return the HDateInfo for the upcoming or current Erev Yom Tov."""
        if self.next_day.is_yom_tov:
            return self
        if self.is_yom_tov:  # Tomorrow is not yom tov, get next erev yom tov
            return self.next_day.upcoming_erev_yom_tov
        return self.upcoming_yom_tov.previous_day

    @property
    def upcoming_shabbat_or_yom_tov(self) -> HDateInfo:
        """Return the HDateInfo for the upcoming or current Shabbat or Yom Tov."""
        if self.is_shabbat or self.is_yom_tov:
            return self

        if self.upcoming_yom_tov.gdate < self.upcoming_shabbat.gdate:
            return self.upcoming_yom_tov
        return self.upcoming_shabbat

    @property
    def upcoming_erev_shabbat_or_erev_yom_tov(self) -> HDateInfo:
        """Return the HDateInfo for upcoming or current Erev Shabbat or Erev Yom Tov."""
        return self.upcoming_shabbat_or_yom_tov.previous_day

    @property
    def first_day(self) -> HDateInfo:
        """Return the first day of Yom Tov or Shabbat.

        This is useful for three-day holidays, for example: it will return the
        first in a string of Yom Tov + Shabbat.
        If this HDateInfo is Shabbat followed by no Yom Tov, returns the Saturday.
        If this HDateInfo is neither Yom Tov, nor Shabbat, this just returns
        itself.
        """
        day_iter = self
        while day_iter.previous_day.is_yom_tov or day_iter.previous_day.is_shabbat:
            day_iter = day_iter.previous_day
        return day_iter

    @property
    def last_day(self) -> HDateInfo:
        """Return the last day of Yom Tov or Shabbat.

        This is useful for three-day holidays, for example: it will return the
        last in a string of Yom Tov + Shabbat.
        If this HDate is Shabbat followed by no Yom Tov, returns the Saturday.
        If this HDate is neither Yom Tov, nor Shabbat, this just returns
        itself.
        """
        day_iter = self
        while day_iter.next_day.is_yom_tov or day_iter.next_day.is_shabbat:
            day_iter = day_iter.next_day
        return day_iter
