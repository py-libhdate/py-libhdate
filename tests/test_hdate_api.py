# -*- coding: utf-8 -*-

"""
These tests are based on the API calls made to hdate by homeassistant (and
maybe other apps in the future).
"""

from __future__ import print_function

from datetime import date, datetime

from hdate import HDate, Location, Zmanim


class TestHDateAPI(object):
    def test_readme_example_english(self, capsys):
        test_date = date(2016, 4, 18)
        hdate = HDate(test_date, hebrew=False)
        print(hdate)
        captured = capsys.readouterr()
        assert captured.out == "Monday 10 Nisan 5776\n"

    def test_readme_example_hebrew(self, capsys):
        test_date = date(2016, 4, 26)
        hdate = HDate(test_date, hebrew=True)
        print(hdate)
        captured = capsys.readouterr()
        assert (
            captured.out == u"יום שלישי י\"ח בניסן ה' תשע\"ו ג' בעומר חול המועד פסח\n"
        )

    def test_get_hebrew_date(self):
        """Print the hebrew date."""
        test_date = datetime(2018, 11, 2)
        assert HDate(test_date).hebrew_date == u'כ"ד מרחשוון ה\' תשע"ט'
        assert HDate(test_date, hebrew=False).hebrew_date == "24 Marcheshvan 5779"

    def test_get_upcoming_parasha(self):
        """Check that the upcoming parasha is correct."""
        test_date = datetime(2018, 11, 2)
        assert HDate(test_date).parasha == u"חיי שרה"
        assert HDate(test_date, hebrew=False).parasha == "Chayei Sara"

    def test_get_upcoming_parasha_vezot_habracha(self):
        """Check that the upcoming parasha is correct for vezot habracha."""
        test_date = datetime(2018, 9, 30)
        assert HDate(test_date).parasha == u"וזאת הברכה"
        assert HDate(test_date, hebrew=False).parasha == "Vezot Habracha"

    def test_get_holiday_description(self):
        """Check that the holiday description is correct."""
        test_date = datetime(2018, 12, 3)
        assert HDate(test_date).holiday_description == u"חנוכה"
        assert HDate(test_date, hebrew=False).holiday_description == "Chanukah"


class TestZmanimAPI(object):
    def test_readme_example_hebrew(self, capsys):
        c = Location("פתח תקוה", 32.08707, 34.88747, "Asia/Jerusalem", 54)
        z = Zmanim(date=date(2016, 4, 18), location=c)
        print(z)
        captured = capsys.readouterr()
        assert (
            captured.out == u"עלות השחר - 04:53:00\n"
            u"זמן טלית ותפילין - 05:19:00\n"
            u"הנץ החמה - 06:09:00\n"
            u'סוף זמן ק"ש מג"א - 08:46:00\n'
            u'סוף זמן ק"ש גר"א - 09:24:00\n'
            u'סוף זמן תפילה מג"א - 10:03:40\n'
            u'סוף זמן תפילה גר"א - 10:29:00\n'
            u"חצות היום - 12:39:00\n"
            u"מנחה גדולה - 13:11:30\n"
            u"מנחה קטנה - 16:26:30\n"
            u"פלג המנחה - 17:48:45\n"
            u"שקיעה - 19:10:00\n"
            u"צאת הכוכבים - 19:35:00\n"
            u"חצות הלילה - 00:39:00\n\n"
        )

    def test_readme_example_english(self, capsys):
        c = Location("פתח תקוה", 32.08707, 34.88747, "Asia/Jerusalem", 54)
        z = Zmanim(date=date(2016, 4, 18), location=c, hebrew=False)
        print(z)
        captured = capsys.readouterr()
        assert (
            captured.out == "Alot HaShachar - 04:53:00\n"
            "Talit & Tefilin's time - 05:19:00\n"
            "Sunrise - 06:09:00\n"
            'Shema EOT MG"A - 08:46:00\n'
            'Shema EOT GR"A - 09:24:00\n'
            'Tefila EOT MG"A - 10:03:40\n'
            'Tefila EOT GR"A - 10:29:00\n'
            "Midday - 12:39:00\n"
            "Big Mincha - 13:11:30\n"
            "Small Mincha - 16:26:30\n"
            "Plag Mincha - 17:48:45\n"
            "Sunset - 19:10:00\n"
            "First stars - 19:35:00\n"
            "Midnight - 00:39:00\n\n"
        )

    def test_issur_melacha_weekday(self):
        c = Location("פתח תקוה", 32.08707, 34.88747, "Asia/Jerusalem", 54)
        z = Zmanim(date=date(2018, 11, 12), location=c)
        assert not z.issur_melacha_in_effect

    def test_issur_melacha_shabbat_morning(self):
        c = Location("פתח תקוה", 32.08707, 34.88747, "Asia/Jerusalem", 54)
        z = Zmanim(date=datetime(2018, 11, 10, 9), location=c)
        assert z.issur_melacha_in_effect

    def test_issur_melacha_friday_morning(self):
        c = Location("פתח תקוה", 32.08707, 34.88747, "Asia/Jerusalem", 54)
        z = Zmanim(date=datetime(2018, 11, 9, 9, 45), location=c)
        assert not z.issur_melacha_in_effect

    def test_issur_melacha_friday_evening(self):
        c = Location("פתח תקוה", 32.08707, 34.88747, "Asia/Jerusalem", 54)
        z = Zmanim(date=datetime(2018, 11, 9, 16, 45), location=c)
        assert z.issur_melacha_in_effect

    def test_issur_melacha_motsaei_shabbat(self):
        c = Location("פתח תקוה", 32.08707, 34.88747, "Asia/Jerusalem", 54)
        z = Zmanim(date=datetime(2018, 11, 10, 17, 45), location=c)
        assert not z.issur_melacha_in_effect

    def test_issur_melacha_shavuot_morning(self):
        c = Location("פתח תקוה", 32.08707, 34.88747, "Asia/Jerusalem", 54)
        z = Zmanim(date=datetime(2019, 6, 9, 9), location=c)
        assert z.issur_melacha_in_effect

    def test_issur_melacha_pesach_vi_mornng(self):
        c = Location("פתח תקוה", 32.08707, 34.88747, "Asia/Jerusalem", 54)
        z = Zmanim(date=datetime(2019, 4, 25, 9, 45), location=c)
        assert not z.issur_melacha_in_effect

    def test_issur_melacha_shavuot_evening(self):
        c = Location("פתח תקוה", 32.08707, 34.88747, "Asia/Jerusalem", 54)
        z = Zmanim(date=datetime(2019, 6, 8, 21, 45), location=c)
        assert z.issur_melacha_in_effect

    def test_issur_melacha_motsaei_shavuot(self):
        c = Location("פתח תקוה", 32.08707, 34.88747, "Asia/Jerusalem", 54)
        z = Zmanim(date=datetime(2019, 6, 9, 20, 30), location=c)
        assert not z.issur_melacha_in_effect

    def test_issur_melacha_pesach_ii_morning(self):
        c = Location(
            name="New York",
            latitude=40.7128,
            longitude=-74.0060,
            timezone="America/New_York",
            diaspora=True,
        )
        z = Zmanim(date=datetime(2019, 4, 21, 9), location=c)
        assert z.issur_melacha_in_effect

    def test_issur_melacha_pesach_ii_evening(self):
        c = Location(
            name="New York",
            latitude=40.7128,
            longitude=-74.0060,
            timezone="America/New_York",
            diaspora=True,
        )
        z = Zmanim(date=datetime(2019, 4, 20, 21, 45), location=c)
        assert z.issur_melacha_in_effect

    def test_issur_melacha_motsaei_pesach_ii(self):
        c = Location(
            name="New York",
            latitude=40.7128,
            longitude=-74.0060,
            timezone="America/New_York",
            diaspora=True,
        )
        z = Zmanim(date=datetime(2019, 4, 21, 20, 30), location=c)
        assert not z.issur_melacha_in_effect

    def test_zmanim_localized_datetime(self):
        c = Location(
            name="New York",
            latitude=40.7128,
            longitude=-74.0060,
            timezone="America/New_York",
            diaspora=True,
        )
        z = Zmanim(date=c.timezone.localize(datetime(2019, 4, 21, 20, 30)), location=c)
        assert not z.issur_melacha_in_effect
