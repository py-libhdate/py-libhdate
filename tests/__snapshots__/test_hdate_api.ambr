# serializer version: 1
# name: TestZmanimAPI.test_readme_example_english[Petah Tikva]
  '''
  Alot HaShachar - 04:52:00
  Talit & Tefilin's time - 05:18:00
  Sunrise - 06:08:00
  Shema EOT MG"A - 08:46:00
  Shema EOT GR"A - 09:23:00
  Tefila EOT MG"A - 10:04:00
  Tefila EOT GR"A - 10:28:00
  Midday - 12:40:00
  Big Mincha - 13:10:30
  Big Mincha 30 min - 13:10:00
  Small Mincha - 16:25:30
  Plag Mincha - 17:50:45
  Sunset - 19:12:00
  End of fast - 19:40:00
  End of Shabbat - 19:50:00
  Tset Hakochavim (18 minutes) - 19:31:30
  Night by Rabbeinu Tam - 20:30:00
  Midnight - 00:40:00
  '''
# ---
# name: TestZmanimAPI.test_readme_example_hebrew[Petah Tikva]
  '''
  עלות השחר - 04:52:00
  זמן טלית ותפילין - 05:18:00
  הנץ החמה - 06:08:00
  סוף זמן ק"ש מג"א - 08:46:00
  סוף זמן ק"ש גר"א - 09:23:00
  סוף זמן תפילה מג"א - 10:04:00
  סוף זמן תפילה גר"א - 10:28:00
  חצות היום - 12:40:00
  מנחה גדולה - 13:10:30
  מנחה גדולה 30 דק - 13:10:00
  מנחה קטנה - 16:25:30
  פלג המנחה - 17:50:45
  שקיעה - 19:12:00
  מוצאי צום - 19:40:00
  מוצאי שבת - 19:50:00
  צאת הכוכבים (18 דק) - 19:31:30
  לילה לרבנו תם - 20:30:00
  חצות הלילה - 00:40:00
  '''
# ---
